### PR TITLE
Pass document by reference.

### PR DESCRIPTION
**Pull request by tmpfs**
I think it makes more sense to pass a reference to the document when adding to the index as it is only a reference that is passed to the extraction function for each field.

Also I have a use case where I need to store the original documents and this prevents an unnecessary clone when adding to the index.

Thanks 🙏

### CODE REVIEW

**Comment by marcus-pousette** [comment]
Thanks for the PR! 

This looks good. Not sure why I passed documents by value in the first place when creating this.

Before merging: 

- Could you make sure that the ```cargo fmt``` check works as expected? See the failing pipeline step. 

- Could you make sure ```cargo bench``` works? There is a compile error in ```test_benchmark.rs```

**Comment by tmpfs** [comment]
Sure @marcus-pousette, those are fixed now 👍

**Comment by tmpfs** [comment]
Technically this is a breaking change, do you want to bump the major version, maybe `2.0.0-alpha-1`?

**Comment by marcus-pousette** [comment]
A major version bump sounds good. Especially if we could have the improvements in the Filter and Tokenizer as you mentioned. 

Additional things: 

- I realized the readme has to be changed with this PR  See the example where ```add_document_to_index``` is invoked. Should be reference now on the document in stead of values. 

- ```cargo clippy``` is complaining, see failiing pipeline


If you are motivated to make the Tokenizer, and Filter update, feel free to include those in this PR or in another PR

**Comment by tmpfs** [comment]
Thanks @marcus-pousette, lets see if we can get the `Tokenizer` and `Filter` changes into this PR while I am at it, should get to it in the next day or so 👍

**Comment by tmpfs** [comment]
@marcus-pousette, I think this is good now, please review 👍 

The benchmarks show a little performance improvement 😁

File: src/index.rs [hunk 1]
```diff
@@ -290,15 +290,15 @@ pub fn add_document_to_index<T: Eq + Hash + Copy, D>(
     tokenizer: Tokenizer,
     filter: Filter,
     key: T,
-    doc: D,
+    doc: &D,
 ) {
     let docs = &mut index.docs;
     let fields = &mut index.fields;
     let mut field_length = vec![0; fields.len()];
     let mut term_counts: HashMap<String, Vec<usize>> = HashMap::new();
```
Comment (marcus-pousette) [inline src/index.rs:298]: While we are at it, we can make it even more performant: 

```suggestion
    let mut term_counts: HashMap<&str, Vec<usize>> = HashMap::new()
```

File: src/index.rs [hunk 2]
```diff
@@ -308,16 +308,16 @@ pub fn add_document_to_index<T: Eq + Hash + Copy, D>(
...
                     all_terms.push(term.to_owned());
                     filtered_terms_count += 1;
-                    let counts = term_counts.get_mut(&term);
+                    let counts = term_counts.get_mut(term);
                     match counts {
                         None => {
                             let mut new_count = vec![0; fields_len];
                             new_count[i] += 1;
-                            term_counts.insert(term, new_count);
+                            term_counts.insert(term.to_owned(), new_count);
```
Comment (marcus-pousette) [inline src/index.rs:320]: With change above we don't need to make this owned
```suggestion
                            term_counts.insert(term, new_count);
```

File: src/index.rs [hunk 3]
```diff
@@ -308,16 +308,16 @@ pub fn add_document_to_index<T: Eq + Hash + Copy, D>(
             // filter and count terms, ignore empty strings
             let mut filtered_terms_count = 0;
             for mut term in terms {
-                term = filter(&term);
+                term = filter(term);
                 if !term.is_empty() {
                     all_terms.push(term.to_owned());
```
Comment (marcus-pousette) [inline src/index.rs:313]:
```suggestion
                    all_terms.push(term);
```

File: src/index.rs [hunk 4]
```diff
@@ -290,15 +290,15 @@ pub fn add_document_to_index<T: Eq + Hash + Copy, D>(
...
     filter: Filter,
     key: T,
-    doc: D,
+    doc: &D,
 ) {
     let docs = &mut index.docs;
     let fields = &mut index.fields;
     let mut field_length = vec![0; fields.len()];
     let mut term_counts: HashMap<String, Vec<usize>> = HashMap::new();
     let mut all_terms: Vec<String> = Vec::new();
```
Comment (marcus-pousette) [inline src/index.rs:299]: Perhaps this also make sense to include in this PR

```suggestion
    let mut all_terms: Vec<&str> = Vec::new();
```

**Review by marcus-pousette — COMMENTED** [review]
The changes you have done look good. 

I included some suggestions for additional improvements, if you like to apply them for (even more performance gains) feel free to :)

**Comment by tmpfs** [comment]
Good catch @marcus-pousette , I have updated with those changes 👍

**Comment by marcus-pousette** [comment]
Well done! Thanks 👍

**Pull request merged**